### ISSUE TITLE
Definition of 'dot' function

### ISSUE THREAD

**Issue by Alaya-in-Matrix** [alpmestan/regress#1]
Is the `dot` function defined to produce the dot product?

Seems that it does not give dot product

for example `dot [1,2] [1,2]` would give 9 instead of `5 = 1 * 1 + 2 *2`

Perhaps a workaround like this

``` haskell
dot :: (Applicative v, Foldable v, Num a)
    => v a
    -> v a
    -> a
dot x y = sum $ zipWith (*) xs ys
    where xs = toList x
          ys = toList y
-- dot x y = getSum . foldMap Sum $ liftA2 (*) x y
```

**Comment by alpmestan** [comment]
Right, I have mostly been using vectors where the `Applicative` instance does "the right thing". It's however quite wrong to rely on  Applicative`here. You either have to use`ZipList`or any container that has a "zippy"`Applicative` for the code to behave correctly at the moment. OTOH, it's a bit unfortunate to convert to list in your proposed definition.

Wasn't there a class in some package to express exactly this `zipWith` operation? If not, I could provide one along with well-behaved instances for all the standard containers.

Sorry for the inconvenience, this package was initially a bit of an experiment and hasn't seen a lot of love lately.
